### PR TITLE
refactor: `MergeSortNoExtraSpace`

### PR DESCRIPTION
**Pull request by alxkm**
Refactor MergeSortNoExtraSpace

Changing naming

Adding test

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

File: src/main/java/com/thealgorithms/sorts/MergeSortNoExtraSpace.java [hunk 1]
```diff
@@ -1,73 +1,83 @@
...
+     * Sorts the array using in-place merge sort algorithm.
+     *
+     * @param array the array to be sorted
+     * @return the sorted array
+     */
+    public static int[] sort(int[] array) {
+        if (array.length == 0) {
+            return array;
+        }
+        int maxElement = Arrays.stream(array).max().getAsInt() + 1;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/MergeSortNoExtraSpace.java:23]:
```suggestion
        final int maxElement = Arrays.stream(array).max().getAsInt() + 1;
```

File: src/main/java/com/thealgorithms/sorts/MergeSortNoExtraSpace.java [hunk 2]
```diff
@@ -1,73 +1,83 @@
...
+     * @param end    the ending index of the array
+     * @param maxElement the value greater than any element in the array, used for encoding
+     */
+    public static void mergeSort(int[] array, int start, int end, int maxElement) {
         if (start < end) {
-            int mid = (start + end) / 2;
-            mergeSort(a, start, mid, maxele);
-            mergeSort(a, mid + 1, end, maxele);
-            implementMergeSort(a, start, mid, end, maxele);
+            int middle = (start + end) / 2;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/MergeSortNoExtraSpace.java:38]:
```suggestion
            final int middle = (start + end) >>> 1;
```
I am puzzled that spotbugs, does not trigger the `IM_AVERAGE_COMPUTATION_COULD_OVERFLOW` warning here (cf. #5131).

File: src/test/java/com/thealgorithms/sorts/MergeSortNoExtraSpaceTest.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    record TestCase(int[] inputArray, int[] expectedArray) {
+    }
+
+    static Stream<TestCase> provideTestCases() {
+        return Stream.of(new TestCase(new int[] {}, new int[] {}), new TestCase(new int[] {1}, new int[] {1}), new TestCase(new int[] {1, 2, 3, 4, 5}, new int[] {1, 2, 3, 4, 5}), new TestCase(new int[] {5, 4, 3, 2, 1}, new int[] {1, 2, 3, 4, 5}),
+            new TestCase(new int[] {3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5}, new int[] {1, 1, 2, 3, 3, 4, 5, 5, 5, 6, 9}), new TestCase(new int[] {4, 2, 4, 3, 2, 1, 5}, new int[] {1, 2, 2, 3, 4, 4, 5}), new TestCase(new int[] {0, 0, 0, 0}, new int[] {0, 0, 0, 0}),
+            new TestCase(new int[] {1000, 500, 100, 50, 10, 5, 1}, new int[] {1, 5, 10, 50, 100, 500, 1000}), new TestCase(new int[] {1, 2, 3, 1, 2, 3, 1, 2, 3}, new int[] {1, 1, 1, 2, 2, 2, 3, 3, 3}),
+            new TestCase(new int[] {10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0}, new int[] {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}), new TestCase(new int[] {}, new int[] {}), new TestCase(new int[] {1}, new int[] {1}), new TestCase(new int[] {2, 1}, new int[] {1, 2}),
+            new TestCase(new int[] {1, 3, 2}, new int[] {1, 2, 3}));
+    }
```
Comment (vil02) [inline src/test/java/com/thealgorithms/sorts/MergeSortNoExtraSpaceTest.java:19]: Some of the test cases are duplicated:
```suggestion
    static Stream<TestCase> provideTestCases() {
        return Stream.of(new TestCase(new int[] {}, new int[] {}), new TestCase(new int[] {1}, new int[] {1}), new TestCase(new int[] {1, 2, 3, 4, 5}, new int[] {1, 2, 3, 4, 5}), new TestCase(new int[] {5, 4, 3, 2, 1}, new int[] {1, 2, 3, 4, 5}),
            new TestCase(new int[] {3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5}, new int[] {1, 1, 2, 3, 3, 4, 5, 5, 5, 6, 9}), new TestCase(new int[] {4, 2, 4, 3, 2, 1, 5}, new int[] {1, 2, 2, 3, 4, 4, 5}), new TestCase(new int[] {0, 0, 0, 0}, new int[] {0, 0, 0, 0}),
            new TestCase(new int[] {1000, 500, 100, 50, 10, 5, 1}, new int[] {1, 5, 10, 50, 100, 500, 1000}), new TestCase(new int[] {1, 2, 3, 1, 2, 3, 1, 2, 3}, new int[] {1, 1, 1, 2, 2, 2, 3, 3, 3}),
            new TestCase(new int[] {10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0}, new int[] {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}), new TestCase(new int[] {2, 1}, new int[] {1, 2}), new TestCase(new int[] {1, 3, 2}, new int[] {1, 2, 3}));
    }
```

File: src/main/java/com/thealgorithms/sorts/MergeSortNoExtraSpace.java [hunk 4]
```diff
@@ -1,73 +1,83 @@
...
-    public static void callMergeSort(int[] a, int n) {
-        int maxele = Arrays.stream(a).max().getAsInt() + 1;
-        mergeSort(a, 0, n - 1, maxele);
+    /**
+     * Sorts the array using in-place merge sort algorithm.
+     *
+     * @param array the array to be sorted
+     * @return the sorted array
+     */
+    public static int[] sort(int[] array) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/sorts/MergeSortNoExtraSpace.java:19]: This implementation does not handle negative inputs: is that right? If yes, please throw `IllegalArgumentException` in such situation.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/sorts/MergeSortNoExtraSpace.java (hunk 4 above)
Comment (alxkm) [reply src/main/java/com/thealgorithms/sorts/MergeSortNoExtraSpace.java:19]: Yes, you are right. Will add this check and test for it.

File: src/main/java/com/thealgorithms/sorts/MergeSortNoExtraSpace.java (hunk 2 above)
Comment (alxkm) [reply src/main/java/com/thealgorithms/sorts/MergeSortNoExtraSpace.java:38]: Really good approach, thanks.

**Review by vil02 — APPROVED** [review]

**Pull request merged**